File: js/agents/src/patterns/blackboard.ts
import type { AgentRoom } from "../AgentRoom";
import type { StateEnvelope } from "../types";
import { createStateEnvelope } from "../envelope";

/**
 * Blackboard pattern — shared state across agents.
 *
 * Agents read and write key-value pairs visible to all room participants.
 * Uses retained messages so state is available on join.
 */
export class Blackboard {
  private _room: AgentRoom;
  private _agentId: string;
  private _state = new Map<string, StateEnvelope>();

  constructor(room: AgentRoom, agentId: string) {
    this._room = room;
    this._agentId = agentId;

    this._room.on("stateChange", (envelope) => {
      this._state.set(envelope.key, envelope);
    });
  }

  /**
   * Set a shared state value.
   */
  set(key: string, value: unknown): void {
    const existing = this._state.get(key);
    const version = existing ? existing.version + 1 : 1;
    const envelope = createStateEnvelope(key, value, version, this._agentId);
    this._state.set(key, envelope);
    this._room.publishState(envelope as unknown as Record<string, unknown>);
  }

  /**
   * Get a shared state value.
   */
  get(key: string): unknown | undefined {
    return this._state.get(key)?.value;
  }

  /**
   * Get the full state envelope for a key.
   */
  getEnvelope(key: string): StateEnvelope | undefined {
    return this._state.get(key);
  }

  /**
   * Get all state entries.
   */
  getAll(): ReadonlyMap<string, StateEnvelope> {
    return this._state;
  }

  /**
   * Register a handler for state changes on a specific key.
   */
  onChange(
    key: string,
    handler: (envelope: StateEnvelope) => void,
  ): void {
    this._room.on("stateChange", (envelope) => {
      if (envelope.key === key) {
        handler(envelope);
      }
    });
  }
}

File: python/agents/nolag_agents/patterns/blackboard.py
from __future__ import annotations

from typing import Any, Callable, Optional

from ..agent_room import AgentRoom
from ..types import StateEnvelope
from ..envelope import create_state_envelope


class Blackboard:
    """Shared state across agents with versioning.

    Agents read and write key-value pairs visible to all room participants.
    Uses retained messages so state is available on join.
    """

    def __init__(self, room: AgentRoom, agent_id: str) -> None:
        self._room = room
        self._agent_id = agent_id
        self._state: dict[str, StateEnvelope] = {}

        self._room.on("state_change", self._on_state_change)

    def _on_state_change(self, data: Any) -> None:
        d = data if isinstance(data, dict) else {}
        envelope = _dict_to_state(d)
        self._state[envelope.key] = envelope

    async def set(self, key: str, value: Any) -> None:
        existing = self._state.get(key)
        version = (existing.version + 1) if existing else 1
        envelope = create_state_envelope(key, value, version, self._agent_id)
        self._state[key] = envelope
        await self._room.publish_state(envelope.to_dict())

    def get(self, key: str) -> Any:
        entry = self._state.get(key)
        return entry.value if entry else None

    def get_envelope(self, key: str) -> Optional[StateEnvelope]:
        return self._state.get(key)

    def get_all(self) -> dict[str, StateEnvelope]:
        return dict(self._state)

    def on_change(self, key: str, handler: Callable[[StateEnvelope], None]) -> None:
        def _handler(data: Any) -> None:
            d = data if isinstance(data, dict) else {}
            envelope = _dict_to_state(d)
            if envelope.key == key:
                handler(envelope)

        self._room.on("state_change", _handler)


def _dict_to_state(d: dict[str, Any]) -> StateEnvelope:
    return StateEnvelope(
        type="state",
        key=d.get("key", ""),
        value=d.get("value"),
        version=d.get("version", 0),
        updated_at=d.get("updatedAt", d.get("updated_at", 0)),
        updated_by=d.get("updatedBy", d.get("updated_by", "")),
    )
